### PR TITLE
feat(eslint-plugin): renamed exported recommended

### LINKED ISSUE #1487: [Feature]: Split the linter plugin to multiple plugins
### Context

Today Otter exposes one single plugin with rules dedicated to different features.
In the current setup, all the Otter Linter rules are per default activated which require additional dependency to an application that would not use it in its own code.

### Proposal

- [ ] Split the code of the plugin to the dedicated package
- [ ] Expose the linter plugin and it's recommended configuration as subentries
- [ ] The additional required dependencies should be peer optional dependencies
- [ ] A package to share helpers may need to be created
- [ ] The `@o3r/eslint-config-otter` package should setup the rules of the modules provided by the application (using `import()` ?).
- [ ] An `ngUpdate` should be provided to `@o3r/eslint-plugin` to migrate to the import of the different packages
- [ ] `@o3r/eslint-plugin` should be deprecated

### LINKED ISSUE #2482: [Bug]: several peerDeps are flagged as optional in @o3r/eslint-plugin but usage of it are exposed at the root
### Package name

eslint-plugin

### Package version

11.4.1

### Reproduction steps

Use `@o3r/eslint-plugin` without having `yaml-eslint-parser` installed.

### Current result

Issue while running linter because `yaml-eslint-parser` is not installed.

### Expected result

To not introduce breaking changes, in v11 we can remove the optional flags.
For v12 we can expose sub-entries like `@o3r/eslint-plugin/yaml`

### Additional comments

Same issue for `jsonc-eslint-parser` and `@angular-eslint/template-parser`

### PR DESCRIPTION
**Pull request by kpanot**
## Proposed change

feat(eslint-plugin): renamed exported recommended
fix(eslint-plugin): turn yaml parser really optional
deprecate(eslint-plugin): recommended exports based on types

## Related issues

- :rocket: Feature resolves #1487
- 🐛 Fix resolves #2482

### CODE REVIEW

File: packages/@o3r/eslint-plugin/src/index.ts [hunk 1]
```diff
@@ -39,19 +39,34 @@ module.exports = {
...
+    // deprecated: should use `monorepo-recommended` instead
     'json-recommended': {
       rules: {
         '@o3r/json-dependency-versions-harmonize': 'error'
       }
     },
 
+    'angular-template-recommended': {
+      rules: {
+        '@o3r/no-folder-import-for-module': 'error',
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/src/index.ts:59]: This tule as nothing to do in `template` rules

File: packages/@o3r/eslint-plugin/src/index.ts [hunk 2]
```diff
@@ -39,19 +39,34 @@ module.exports = {
...
     },
 
+    // deprecated: should use `monorepo-recommended` instead
     'json-recommended': {
       rules: {
         '@o3r/json-dependency-versions-harmonize': 'error'
       }
     },
 
+    'angular-template-recommended': {
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/src/index.ts:57]: This config contains o3r rules 
why renaming it with `angular` prefix ?

File: packages/@o3r/eslint-plugin/schematics/ng-update/v11/update-configs/update-configs.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import type { Rule } from '@angular-devkit/schematics';
+import { basename } from 'node:path';
+
+/**
+ * Update Otter cms.json in an Angular Project
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/schematics/ng-update/v11/update-configs/update-configs.ts:5]: description does not match

File: packages/@o3r/eslint-plugin/schematics/ng-update/v11/index.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    const updateRules: Rule[] = [
+      updateEslintJsonFile()
+    ];
+
+    return chain(updateRules)(tree, context);
+  };
+}
+
+/**
+ * Update of Otter library V10.0
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/schematics/ng-update/v11/index.ts:20]: description :)

File: packages/@o3r/eslint-plugin/schematics/ng-update/v11/index.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+/* eslint-disable camelcase, @typescript-eslint/naming-convention */
+import { chain, Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
+import { updateEslintJsonFile } from './update-configs/update-configs';
+
+/**
+ * Update of Otter library V10.0
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/schematics/ng-update/v11/index.ts:6]: here as well :)

File: packages/@o3r/eslint-plugin/schematics/ng-update/v11/update-configs/update-configs.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  const configFilePattern = /^.?eslint(:?rc)?\..*/;
+  const toReplace = {
+    // eslint-disable-next-line @typescript-eslint/naming-convention -- following eslint config name
+    'angular-template-recommended': /(['"](?:plugin:)?@o3r\/)template-recommended(['"])/g,
+    // eslint-disable-next-line @typescript-eslint/naming-convention -- following eslint config name
+    'monorepo-recommended': /(['"](?:plugin:)?@o3r\/)json-recommended(['"])/g
+  };
+
+  return (tree) => {
+    tree.visit((path) => {
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/schematics/ng-update/v11/update-configs/update-configs.ts:17]: We may not use `tree.visit` because it visit all node_modules etc...
we have an helper in @o3r/schematics `findFilesInTree` that is not visiting these folders

File: packages/@o3r/eslint-plugin/migration.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
+  "$schema": "https://raw.githubusercontent.com/angular/angular-cli/master/packages/angular_devkit/schematics/collection-schema.json",
+  "schematics": {
+    "migration-v11_4": {
+      "version": "11.4.0-alpha.0",
+      "description": "Updates of @o3r/eslint-plugin to v11.4.*",
+      "factory": "./schematics/ng-update/v10-0/index#updateV11_4"
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/migration.json:7]:
```suggestion
      "factory": "./schematics/ng-update/v11/index#updateV11_4"
```

**Review by matthieu-crouzet — COMMENTED** [review]
Shouldn't we do the same ng-update for the `@o3r/eslint-config-otter` 
as the ng-add of `@o3r/eslint-plugin` does nothing but `@o3r/eslint-config-otter` does the setup 
and in `@o3r/eslint-config-otter` we should migrate to the not deprecated configs



Moreover do we really need this fix right now as we are working on the ESLint flat config  ? or could it be done directly inside the branch for this migration to flat config ?

File: packages/@o3r/eslint-plugin/src/index.ts (hunk 1 above)
Comment (kpanot) [reply packages/@o3r/eslint-plugin/src/index.ts:59]: It is then wrongly part of `template-recommended`, I will remove it

File: packages/@o3r/eslint-plugin/src/index.ts (hunk 2 above)
Comment (kpanot) [reply packages/@o3r/eslint-plugin/src/index.ts:57]: The `@o3r/` prefix remains, but this are recommended for angular template parsing only.
If you see another name, no problem to propose it :)

File: packages/@o3r/eslint-plugin/src/index.ts [hunk 8]
```diff
@@ -39,19 +39,34 @@ module.exports = {
       }
     },
 
+    // deprecated: should use `angular-template-recommended` instead
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/src/index.ts:42]: We may specify the version it will become deprecated

File: packages/@o3r/eslint-plugin/src/index.ts (hunk 2 above)
Comment (matthieu-crouzet) [reply packages/@o3r/eslint-plugin/src/index.ts:57]: Does the no-inner-html rely on angular ? 
the name doesn't seem to imply this

File: packages/@o3r/eslint-plugin/src/index.ts [hunk 9]
```diff
@@ -39,19 +39,34 @@ module.exports = {
...
     'json-recommended': {
       rules: {
         '@o3r/json-dependency-versions-harmonize': 'error'
       }
     },
 
+    'angular-template-recommended': {
+      rules: {
+        '@o3r/no-inner-html': 'off',
+        '@o3r/template-async-number-limitation': 'warn'
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-plugin/src/index.ts:60]: Our new recommendation is to have 'error' or 'off' by default

File: packages/@o3r/eslint-config-otter/schematics/ng-update/v11.4/update-configs/update-configs.spec.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { Tree } from '@angular-devkit/schematics';
+import { updateEslintRecommended } from './update-configs';
+
+describe('updateEslintRecommended', () => {
+  it('should update configs', async () => {
+    const initialTree = Tree.empty();
+    initialTree.create('random.file', 'a file containing json-recommended');
+    initialTree.create('/test/.eslintrc.json', '{ "extends": ["@o3r/json-recommended", "@other/json-recommended"] }');
+    initialTree.create('/src/.eslintrc.json', '{ "extends": ["plugin:@o3r/json-recommended", "@o3r/template-recommended"] }');
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-config-otter/schematics/ng-update/v11.4/update-configs/update-configs.spec.ts:9]: You don't have test for `eslint.config.{m,c,}js` that match the regexp `/^.?eslint(:?rc)?\..*/`

File: packages/@o3r/eslint-plugin/src/index.ts (hunk 2 above)
Comment (kpanot) [reply packages/@o3r/eslint-plugin/src/index.ts:57]: It forbids `[innterHtml]` directive in an Angular Template.
(It also requires the `@angular-eslint/template-parser`)

File: packages/@o3r/eslint-config-otter/migration.json [hunk 11]
```diff
@@ -5,6 +5,11 @@
       "version": "10.0.0-alpha.0",
       "description": "Updates of @o3r/eslint-config-otter to v10.0.*",
       "factory": "./schematics/ng-update/index#updateV100"
+    },
+    "migration-v11_4": {
```
Comment (vscaiceanu-1a) [inline packages/@o3r/eslint-config-otter/migration.json:9]: maybe you want to target the `release/11.4.0-rc` branch instead of `main`

File: packages/@o3r/eslint-plugin/src/index.ts [hunk 12]
```diff
@@ -39,19 +39,34 @@ module.exports = {
...
+    // deprecated: should use `monorepo-recommended` instead. Will be removed in v13.
     'json-recommended': {
       rules: {
         '@o3r/json-dependency-versions-harmonize': 'error'
       }
     },
 
+    'angular-template-recommended': {
+      rules: {
+        '@o3r/no-inner-html': 'off',
```
[the remark this answers is not in the document]
Comment (kpanot) [reply packages/@o3r/eslint-plugin/src/index.ts:59]: It is the current one.
It can be changed, in the big eslint refactor if we want

File: packages/@o3r/eslint-config-otter/migration.json (hunk 11 above)
Comment (kpanot) [reply packages/@o3r/eslint-config-otter/migration.json]: nop `main` is the right target but the PR was created before the `rc`.
I changed for `11.5` instead

File: packages/@o3r/eslint-config-otter/schematics/ng-update/index.ts [hunk 13]
```diff
@@ -16,7 +17,26 @@ function updateV100fn(): Rule {
...
+
 /**
  * Update of Otter library V10.0
  */
 export const updateV100 = createSchematicWithMetricsIfInstalled(updateV100fn);
+
+/**
+ * Update of Otter library V11.5
+ */
+export const updateV114 = createSchematicWithMetricsIfInstalled(updateV115Fn);
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-config-otter/schematics/ng-update/index.ts:42]:
```suggestion
export const updateV115 = createSchematicWithMetricsIfInstalled(updateV115Fn);
```

**Review by matthieu-crouzet — APPROVED** [review]

File: codecov.yml [hunk 14]
```diff
@@ -1,5 +1,14 @@
 codecov:
   require_ci_to_pass: true
+  max_report_age: off
```
Comment (matthieu-crouzet) [inline codecov.yml:3]: This should not be part of this PR

File: packages/@o3r/eslint-config-otter/migration.json [hunk 15]
```diff
@@ -5,6 +5,11 @@
       "version": "10.0.0-alpha.0",
       "description": "Updates of @o3r/eslint-config-otter to v10.0.*",
       "factory": "./schematics/ng-update/index#updateV100"
+    },
+    "migration-v11_5": {
+      "version": "11.5.0-alpha.0",
+      "description": "Updates of @o3r/eslint-config-otter to v11.5.*",
+      "factory": "./schematics/ng-update/index#updateV114"
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-config-otter/migration.json:12]:
```suggestion
    "migration-v11_6": {
      "version": "11.6.0-alpha.0",
      "description": "Updates of @o3r/eslint-config-otter to v11.6.*",
      "factory": "./schematics/ng-update/index#updateV116"
```

File: packages/@o3r/eslint-config-otter/schematics/ng-update/index.ts [hunk 16]
```diff
@@ -22,7 +25,25 @@ function updateV100fn(): Rule {
   };
 }
 
+/**
+ * Update of Otter library V11.5
+ */
+function updateV115Fn(): Rule {
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-config-otter/schematics/ng-update/index.ts:31]:
```suggestion
 * Update of Otter library V11.6
 */
function updateV116Fn(): Rule {
```

File: packages/@o3r/eslint-config-otter/schematics/ng-update/index.ts [hunk 17]
```diff
@@ -8,6 +8,9 @@ import {
 import {
   addStylistic,
 } from './v10.0/stylistic';
+import {
+  updateEslintRecommended,
+} from './v11.5/update-configs/update-configs';
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-config-otter/schematics/ng-update/index.ts:13]:
```suggestion
} from './v11.6/update-configs/update-configs';
```
folder should be renamed

File: packages/@o3r/eslint-config-otter/schematics/ng-update/index.ts [hunk 18]
```diff
@@ -22,7 +25,25 @@ function updateV100fn(): Rule {
...
+
 /**
  * Update of Otter library V10.0
  */
 export const updateV100 = createSchematicWithMetricsIfInstalled(updateV100fn);
+
+/**
+ * Update of Otter library V11.5
+ */
+export const updateV115 = createSchematicWithMetricsIfInstalled(updateV115Fn);
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-config-otter/schematics/ng-update/index.ts:49]:
```suggestion
 * Update of Otter library V11.6
 */
export const updateV116 = createSchematicWithMetricsIfInstalled(updateV116Fn);
```

File: packages/@o3r/eslint-plugin/src/index.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (matthieu-crouzet) [reply packages/@o3r/eslint-plugin/src/index.ts:58]: We don't want to prevent the usage of innerHTML, this rule as been done for project that prefer innerText 
In fact we should rename that rule to `@o3r/prefer-inner-text` could be done on the next major but we will still keep this rule out of recommended

File: packages/@o3r/eslint-config-otter/migration.json [hunk 19]
```diff
@@ -5,6 +5,11 @@
       "version": "10.0.0-alpha.0",
       "description": "Updates of @o3r/eslint-config-otter to v10.0.*",
       "factory": "./schematics/ng-update/index#updateV100"
+    },
+    "migration-v11_6": {
+      "version": "11.6.0-alpha.0",
```
Comment (fpaul-1A) [inline packages/@o3r/eslint-config-otter/migration.json:10]: not sure it matters, but we could update this to `prerelease`

File: packages/@o3r/eslint-plugin/src/rules/yaml/yarnrc-package-extensions-harmonize/yarnrc-package-extensions-harmonize.ts [hunk 20]
```diff
@@ -99,49 +106,48 @@ export default createRule<[YarnrcPackageExtensionsHarmonizeOptions, ...any], 've
...
-                    messageId: 'versionUpdate',
-                    data: {
-                      version
-                    },
-                    fix: (fixer) => fixer.replaceTextRange(node.value!.range, `${version}`)
-                  }
-                ]
-              });
+      return {
+        // eslint-disable-next-line @typescript-eslint/naming-convention -- name required by Yaml Eslint parser
```
Comment (fpaul-1A) [inline packages/@o3r/eslint-plugin/src/rules/yaml/yarnrc-package-extensions-harmonize/yarnrc-package-extensions-harmonize.ts:110]: do we still need that?

File: codecov.yml (hunk 14 above)
Comment (kpanot) [reply codecov.yml:3]: Splitted out in #2529

File: packages/@o3r/eslint-plugin/src/rules/yaml/yarnrc-package-extensions-harmonize/yarnrc-package-extensions-harmonize.ts (hunk 20 above)
Comment (kpanot) [reply packages/@o3r/eslint-plugin/src/rules/yaml/yarnrc-package-extensions-harmonize/yarnrc-package-extensions-harmonize.ts:110]: Yes, it was still required
(I think we have a linter for the unneeded linter rules)

Comment (matthieu-crouzet) [reply packages/@o3r/eslint-plugin/src/rules/yaml/yarnrc-package-extensions-harmonize/yarnrc-package-extensions-harmonize.ts:110]: Yes we have one linter option to detect un-needed disable

File: packages/@o3r/eslint-config-otter/schematics/ng-update/v11.6/update-configs/update-configs.spec.ts [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+import {
+  updateEslintRecommended,
+} from './update-configs';
+
+describe('updateEslintRecommended', () => {
+  it('should update configs', async () => {
+    const initialTree = Tree.empty();
+    initialTree.create('random.file', 'a file containing json-recommended');
+    initialTree.create('/test/.eslintrc.json', '{ "extends": ["@o3r/json-recommended", "@other/json-recommended"] }');
+    initialTree.create('/src/eslint.config.mjs', 'export { "extends": ["plugin:@o3r/json-recommended", "@o3r/template-recommended"] };');
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-config-otter/schematics/ng-update/v11.6/update-configs/update-configs.spec.ts:13]: config with mjs does not work with extends 
this package is supposed to be used only with eslintrc config 
so we don't need to validate that no ?

**Review by fpaul-1A — DISMISSED** [review]

**Review by matthieu-crouzet — APPROVED** [review]

**Review by mrednic-1A — APPROVED** [review]

**Review by fpaul-1A — APPROVED** [review]

**Review by mrednic-1A — APPROVED** [review]
